Sources/Apollo/URLSessionClient.swift

```diff
@@ -247,21 +253,47 @@ open class URLSessionClient: NSObject, URLSessionDelegate, URLSessionTaskDelegat
   
   // MARK: - URLSessionDataDelegate
   
-  open func urlSession(_ session: URLSession,
-                       dataTask: URLSessionDataTask,
-                       didReceive data: Data) {
+  open func urlSession(
+    _ session: URLSession,
+    dataTask: URLSessionDataTask,
+    didReceive data: Data
+  ) {
     guard dataTask.state != .canceling else {
       // Task is in the process of cancelling, don't bother handling its data.
       return
     }
-    
-    self.$tasks.mutate {
-      guard let taskData = $0[dataTask.taskIdentifier] else {
-        assertionFailure("No data found for task \(dataTask.taskIdentifier), cannot append received data")
+
+    guard let taskData = self.tasks[dataTask.taskIdentifier] else {
+      assertionFailure("No data found for task \(dataTask.taskIdentifier), cannot append received data")
+      return
+    }
+
+    taskData.append(additionalData: data)
+
+    if let httpResponse = dataTask.response as? HTTPURLResponse, httpResponse.isMultipart {
```

Comment:
This multipart boundary check is to ensure that only complete chunks are sent to the next interceptor. Any remaining data that was received that should be processed with the next incoming data to form a complete chunk will get held back.

Reply:
Will that chunk get automatically included in the `dataTask` next time this function is called? I'm not seeing anywhere that you are holding onto the `dataTask` to merge them together when the next chunk comes in.

Reply:
All incoming data is appended to any data already related to the task, see line [271](https://github.com/apollographql/apollo-ios/pull/2870/files/2de76f1a53ec80fb1d56107b2852f9a010e9c41e#diff-80dc6761af321b52c5b1108a063a8ad3f0086925f9a813c6eb1fe043ada83c86R271). Then we work with the data sending off the complete chunks to other interceptors and on line [290](https://github.com/apollographql/apollo-ios/pull/2870/files/2de76f1a53ec80fb1d56107b2852f9a010e9c41e#diff-80dc6761af321b52c5b1108a063a8ad3f0086925f9a813c6eb1fe043ada83c86R290) is where the task data is reset with any remaining bytes, ready for the callback block of data to be added to it.

Reply:
Got it. To me, this makes it even more important that task data is mutated atomically then. Totally good with this implementation in every other regard. But seems very easy for this to get corrupted by parallel threads executing in a weird order.

Reply:
👍🏻 - created https://github.com/apollographql/apollo-ios/issues/2900 to handle it.